drivers/fuel_gauge/axp2101/fuel_gauge_axp2101.c

```diff
@@ -0,0 +1,204 @@
+	struct i2c_dt_spec i2c;
+};
+
+struct axp2101_data {
+	uint8_t features;
+};
+
+static int enable_fuel_gauge(const struct device *dev);
+static int enable_batt_detection(const struct device *dev);
+
+static int axp2101_init(const struct device *dev)
+{
+	const struct axp2101_config *cfg;
+	struct axp2101_data *data = dev->data;
+	int ret = 0;
+
+	cfg = dev->config;
+
+	if (!device_is_ready(cfg->i2c.bus)) {
+		LOG_ERR("Bus device is not ready");
+		return -ENODEV;
+	}
+
+	ret = enable_fuel_gauge(dev);
+	if (ret < 0) {
+		LOG_WRN("Failed to enable fuel gauge");
+		data->features &= ~GAUGE_FEATURE_GAUGE;
+	}
+
+	ret = enable_batt_detection(dev);
+	if (ret < 0) {
+		LOG_WRN("Failed to enable battery detection");
+		data->features &= ~GAUGE_FEATURE_BAT_DET;
+	}
+	return 0;
+}
+
+static int enable_fuel_gauge(const struct device *dev)
+{
+	int ret = 0;
+	const struct axp2101_config *cfg = dev->config;
+	struct axp2101_data *data = dev->data;
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, AXP2101_CHARGE_GAUGE_WDT_CTRL, GAUGE_ENABLE_MASK,
+				     GAUGE_ENABLE_MASK);
+	if (ret < 0) {
+		data->features &= ~GAUGE_FEATURE_GAUGE;
+	}
+	return ret;
+}
+
+static int enable_batt_detection(const struct device *dev)
+{
+	int ret = 0;
+	const struct axp2101_config *cfg = dev->config;
+	struct axp2101_data *data = dev->data;
+	ret = i2c_reg_update_byte_dt(&cfg->i2c, AXP2101_BAT_DET_CTRL, BAT_TYPE_DET_MASK,
+				     BAT_TYPE_DET_MASK);
+	if (ret < 0) {
+		data->features &= ~GAUGE_FEATURE_BAT_DET;
+	}
+	return ret;
+}
+
+static int is_battery_connect(const struct device *dev, union fuel_gauge_prop_val *val)
+{
+	uint8_t tmp;
+	int ret;
+	const struct axp2101_config *cfg = dev->config;
+	struct axp2101_data *data = dev->data;
+
+	if ((data->features & GAUGE_FEATURE_BAT_DET) == 0) {
+		return -ENOTSUP;
+	}
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, AXP2101_STATUS1, &tmp);
+	if (ret < 0) {
+		return ret;
+	}
+
+	val->present_state = tmp & BAT_PRESENT_MASK;
+	return 0;
+}
+
+static int get_battery_percent(const struct device *dev, union fuel_gauge_prop_val *val)
+{
+	uint8_t tmp;
+	int ret;
+	const struct axp2101_config *cfg = dev->config;
+	struct axp2101_data *data = dev->data;
+
+	if ((data->features & GAUGE_FEATURE_GAUGE) == 0) {
+		return -ENOTSUP;
+	}
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, AXP2101_BAT_PERCENT_DATA, &tmp);
+	if (ret < 0) {
+		return ret;
+	}
+
+	val->relative_state_of_charge = tmp;
+	return 0;
+}
+
+static int get_bat_voltage(const struct device *dev, union fuel_gauge_prop_val *val)
+{
+	uint8_t h5, l8;
+	int ret;
+	const struct axp2101_config *cfg = dev->config;
+	struct axp2101_data *data = dev->data;
+
+	if ((data->features & GAUGE_FEATURE_GAUGE) == 0) {
+		return -ENOTSUP;
+	}
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, AXP2101_ADC_DATA_RELUST0, &h5);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_read_byte_dt(&cfg->i2c, AXP2101_ADC_DATA_RELUST1, &l8);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (h5 == -1 || l8 == -1) {
```

Comment:
First of all, you shouldn't compare int to uint32_t.
Also, isn't 255 an acceptable value for either?

Reply:
Oh yes... I took that part from the reference implementation, but the datasheet does not mention anything about it. Now I dropped it, so we just rely on the bus read status.